### ISSUE TITLE
Create small versions of images during build for use in thumbnails

### ISSUE THREAD

**Issue by narthur** [beeminder/blog#313]
```[tasklist]
### Desiderata
- [x] Clarify how this gissue relates to #348
- [x] Add thumbnail generation to the build process?
- [x] See notes from Dan Byler, possibly make new gissues?
```

Currently the new version of the blog does nothing to optimize images. We should bake out smaller versions of blog images during build so we don't serve huge images for list pages, Beeminder homepage, etc.

HT adamwolf (and many users)

## Cognata

* #348

_**Verbata**: images, bandwidth, blog loading speed, page load times,_

**Comment by narthur** [comment]
Astro has an experimental feature for handling this:

https://docs.astro.build/en/guides/assets/

Hopefully that's all we'll need.

**Comment by narthur** [comment]
![noaa-0XUnd0bXqTE-unsplash](https://github.com/beeminder/blog/assets/4655422/0aee4b82-e550-4ac7-8b7a-d942151fa088)

Testing what URLs GitHub uses to serve images.

**Comment by narthur** [comment]
This is done.

**Comment by narthur** [comment]
Vet's PR adds git-level image compression. The change I made for this PR adds build-level image size variant generation. So the two features should work together fine to provide faster images.

**Comment by dreeves** [comment]
Just to make sure this is all wrapped up, here's an old message from a user with some ideas:

> I’m working at the park today, so I’m tethering … and I noticed the images on your blog are loading kind of slowly. If you care about snappy, you could speed that up by optimizing them pre-upload*, hosting in a generic CDN, or a combination of the above. It looks like Astro has some integrations to that effect: https://astro.build/integrations?search=images
> 
> I downloaded a random png image from the blog, threw it into ImageOptim, and got a visually identical image that was ~80% smaller. From my TOOLS- Image compression, hosting.md file, here are some image optimization tools I’ve made note of over the years:
> 
> Local options:
> 
> * ImageOptim
> * Realmac's Squash
> *  [Optimage](https://optimage.app/)
> 
> Web based:
> 
> * [compressor.io](https://compressor.io/)
> * [ShortPixel](https://shortpixel.com/api-tools) - has APIs, Zapier integrations, and more
> 
> Hosting options with dynamic transformations:
> 
> * Cloudinary (used this at AFG for years; no complaints)
> * ImageKit
>   - Now has Google/Amazon-based image tagging and background removal
>   - Sample transformations:
> https://ik.imagekit.io/h7kn17twvu/tr:ot-overlay%20made%20easy,ots-45,otp-20,otbg-FFFFFF/default-image.jpg
> 
> https://ik.imagekit.io/h7kn17twvu/tr:oi-Lightbulb_-pink_kd_GZpFQy0F.png,ox-45,oy-300/default-image.jpg
> 
> https://ik.imagekit.io/h7kn17twvu/tr:oi-Lightbulb_-pink_kd_GZpFQy0F.png,ox-45,oy-300:tr:oi-Money_UzoNNXI63.png,ox-100,oy-5

**Comment by narthur** [comment]
@dreeves As far as future work goes on image optimization it depends on how important build speed is to you. Currently the blog is generating smaller versions of image at build time, which slows down builds. If we switched to something like ImageKit (or cloudflare--I just got an email from them saying they're providing a similar service now), it could speed up our builds significantly, at the complexity cost of integrating with a third-party service.

**Comment by narthur** [comment]
I guess we could also consider finding a way to include generated thumbnails in git using a github action, similar to strickvl's PR.

**Comment by narthur** [comment]
@dreeves When did Dan give you that note? I'm curious if it was before or after I deployed my build-level image resizing.

**Comment by dreeves** [comment]
> When did Dan give you that note? I'm curious if it was before or after I deployed my build-level image resizing.

Aug 24, 2023, 7:21 AM pacific

**Comment by narthur** [comment]
@dreeves Given I don't think we're paying Render.com anything for the blog (?), definitely wouldn't think our current approach is costly.

**Comment by dreeves** [comment]
Good point. The wait when making tweaks for the new build to be live is kinda painful, but normally it's worth the build taking longer. Ooh, idea: can we have two kinds of deploys/builds? Draft mode and prod mode? Like if I'm making tweaks or fussing with css or something i want to minimize build time. Other times i want to do a full build with all the image optimization and stuff. I don't know if this makes sense or is even possible but maybe there's a way to have the best of both worlds somehow?

**Comment by narthur** [comment]
@dreeves Hmm. Not easily possible if you're working directly on the master branch. If you were using separate branches, we could think about changing the settings for preview builds to make them speedier.

**Comment by dreeves** [comment]
What about something in posts.json for site-wide settings? Like {"mode": "prod"} vs {"mode": "draft"} or booleans for whether to do things like optimize images. Or what if there was an explicit list of which posts to build, with a special option for "build all of them"? Then when working on a particular post you could change it to just rebuild that one post. 

(Maybe that last idea doesn't make sense -- it would have to put the prebuilt ones somewhere to somehow retrieve them without building them I guess.)

**Comment by narthur** [comment]
@dreeves Yup, definitely want to avoid introducing options that change build behavior for production builds.

**Comment by narthur** [comment]
@dreeves With merge of #412 Astro's image cache is now preserved between builds, which should effectively eliminate the image optimization time during most builds.

**Comment by dreeves** [comment]
How do we bust the cache when an image changes? Or is that automatic?

**Comment by narthur** [comment]
@dreeves If an image changes that is at the same url it was previously, you'll need to go into Render.com and re-run the deploy without the cache.

**Comment by dreeves** [comment]
Maybe that won't come up often. But what do you mean specifically? Is that a special Render thing, "deploy without cache"?

**Comment by narthur** [comment]
@dreeves Yup! And definitely let me know if this happens more than I thought it would and it becomes annoying.

![image](https://github.com/beeminder/blog/assets/4655422/592760fc-17e0-40d2-b407-954ec0103415)

**Comment by bsoule** [comment]
I noticed that the lastest several blog posts don't have a thumbnail version of the post image. Did something break with generating those? Or like, does it only make thumbnails when you do the "clear build cache & deploy" thing or something?
